libsnark/zk_proof_systems/plonk/srs.hpp

```diff
@@ -179,32 +191,105 @@ template<typename ppT> class plonk_keypair
     plonk_keypair(plonk_keypair<ppT> &&other) = default;
 };
 
-/// Hashes of transcript after prover rounds 1,2,3,4,5
-template<typename ppT> struct transcript_hash_t {
-
-    /// - beta: permutation challenge - hashes of transcript after round 1
-    libff::Fr<ppT> beta;
-    /// - gamma: permutation challenge - hashes of transcript after round 1
-    libff::Fr<ppT> gamma;
-    /// - alpha: quotient challenge - hash of transcript after rounds 1,2
-    libff::Fr<ppT> alpha;
-    /// - zeta: evaluation challenge - hash of transcriptafter rounds 1,2,3
-    libff::Fr<ppT> zeta;
-    /// - nu: opening challenge - hash of transcript after rounds 1,2,3,4
-    /// (denoted by v in [GWC19])
-    libff::Fr<ppT> nu;
-    /// - u: multipoint evaluation challenge -- hash of transcript after rounds
-    /// 1,2,3,4,5
-    libff::Fr<ppT> u;
-
-    /// struct constructor
-    transcript_hash_t(
-        libff::Fr<ppT> &beta,
-        libff::Fr<ppT> &gamma,
-        libff::Fr<ppT> &alpha,
-        libff::Fr<ppT> &zeta,
-        libff::Fr<ppT> &nu,
-        libff::Fr<ppT> &u);
+/// transcript hasher interface
+///
+/// the hasher works in the Prover as follows:
+///
+/// round 1
+///
+///     add_element(buffer, a_eval_exp)
+///     add_element(buffer, b_eval_exp)
+///     add_element(buffer, c_eval_exp)
+///     // buffer = first_output
+///
+/// round 2
+///
+///     beta = hash(str(buffer) + "0")
+///     gamma = hash(str(buffer) + "1")
+///
+///     acc_eval = evaluate_in_exponent(CRS, accumulator_poly.to_coeffs())
+///
+///     add_element(buffer, acc_eval)
+///     // buffer = first_output + second_output
+///
+/// round 3
+///
+///     alpha = hash(str(buffer))
+///
+///     add_element(buffer, t_lo_eval_exp)
+///     add_element(buffer, t_mid_eval_exp)
+///     add_element(buffer, t_hi_eval_exp)
+///     // buffer = first_output + second_output + third_output
+///
+/// round 4
+///
+///     zeta = hash(str(buffer))
+///
+///     add_element(buffer, a_zeta)
+///     add_element(buffer, b_zeta)
+///     add_element(buffer, c_zeta)
+///     add_element(buffer, S_0_zeta)
+///     add_element(buffer, S_1_zeta)
+///     add_element(buffer, accumulator_shift_zeta)
+///     add_element(buffer, t_zeta)
+///     add_element(buffer, r_zeta)
+///     // buffer = first_output + second_output + third_output + fourth_output
+///
+/// round 5
+///
+///     nu = hash(str(buffer))
+///
+///     W_zeta_eval_exp = evaluate_in_exponent(CRS, W_zeta.to_coeffs())
+///     W_zeta_omega_eval_exp = evaluate_in_exponent(CRS,
+///     W_zeta_omega.to_coeffs())
+///
+///     add_element(buffer, W_zeta_eval_exp)
+///     add_element(buffer, W_zeta_omega_eval_exp)
+///
+///     // buffer = first_output + second_output + third_output + fourth_output
+///     + fifth_output
+///
+///     u = hash(str(buffer))
+///
+template<typename ppT> class transcript_hasher
```

Comment:
This needs to be in it's own file(s).  I think we may need to:
1. create a file `/libsnark/transcript_hasher/transcript_hasher.hpp` with just a comment that describes the interface
2. add the `transcript_hasher` as a template parameter to the plonk prover and verifier. 
3. (maybe) create a `transcript_hasher_base` class which can be used as the basis for all implementations.
4. put the dummy transcript hasher (renamed to make it clear it is the implemenation for testing) in the `plonk/tests` directory.  In the tests, the prover and verifier will be instantiated as `plonk_prover<libff::bls12_381_pp, bls12_381_test_vector_transcript_hasher>` 

Later (let's ensure there is an issue for this):
6. create a `transcript_hasher` implementation using SHA3 etc, which can be used on any curve (since it does not need special constants, it can just serialize to binary and feed to SHA3).
7.  Set the default value of the `transcript_hasher` template parameter (on the plonk prover and verifier) to the sha3 hasher.

Reply:
> create a file `/libsnark/transcript_hasher/transcript_hasher.hpp` with just a comment that describes the interface

Just to confirm: file `/libsnark/transcript_hasher/transcript_hasher.hpp` will literally be empty of code and will contain only the transcript_hasher class interface as a comment -- is that your suggestion?

> put the dummy transcript hasher (renamed to make it clear it is the implemenatation for testing) in the `plonk/tests` directory

Do you suggest to move the current transcript hasher implementation from `srs.hpp` and `srs.tcc` to `tests/`? Would yo recommend to have separate files for it or to put it directly in `tests/test_plonk.cpp`?

Reply:
> Just to confirm: file /libsnark/transcript_hasher/transcript_hasher.hpp will literally be empty of code and will contain only the transcript_hasher class interface as a comment -- is that your suggestion?

Yes.  Due to the way C++ works, we can't actually define an interface for it.  This si similar to https://github.com/clearmatics/libff/blob/develop/libff/algebra/curves/public_params.hpp#L30 and https://github.com/clearmatics/libsnark/blob/plonk-transcript-hasher-issue-56/libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp#L22, excep those files include some templated type defs.  There is a better way to do this (will discuss next time), but it's a bit involved.

The file may also contain a base class, if we create one.

> Do you suggest to move the current transcript hasher implementation from srs.hpp and srs.tcc to tests/? Would you recommend to have separate files for it or to put it directly in tests/test_plonk.cpp?

It could be in the test_plonk file, although there is probably already a lot of code in there, in which case better in it's own file.

Reply:
Addressed all steps except 3. The Issue recording the SHA3 hasher (steps 6,7) is assigned #71.

Regarding step 3:

> 3. (maybe) create a transcript_hasher_base class which can be used as the basis for all implementations.

Will this class be located at `/libsnark/transcript_hasher/transcript_hasher.hpp` ?